README.md
# text_switch

A Flutter package for awesome text switch.

## 🚀 Features

1. Customize active and inactive color
2. Customize active and inactive text
3. Text color
4. gradient background
5. border color
6. Supports Null-safety
7. Active subtitle and inactive subtitle
8. Custom height
9. Leading widget for active and inactive text


[![Say Thanks!](https://img.shields.io/badge/Say%20Thanks-!-1EAEDB.svg)](https://saythanks.io/to/sampathnarayanan72%40gmail.com)


## Show some :heart: and star the repo to support the project

 [![GitHub followers](https://img.shields.io/github/followers/narayanansampath?style=social&label=Follow)](https://github.com/narayanansampath)

[![Open Source Love](https://badges.frapsoft.com/os/v1/open-source.svg?v=102)](https://github.com/narayanansampath/text_switch/blob/master/LICENSE)


## 💻 Installation

In the `dependencies:` section of your `pubspec.yaml`, add the following line:

```yaml
text_switch: <latest_version>
```

Import in your project:
```dart
import 'package:text_switch/text_switch.dart';
```

## ❔Basic Usage
```dart
class HomeScreen extends StatefulWidget {
  @override
  _HomeScreenState createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {

  bool isActive = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Custom Switch Example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextSwitch(
              activeText: 'ON',
              inactiveText: 'OFF',
              value: isActive,
              onChanged: (val) {
                print('Value: $val');
                setState(() => isActive = val);
              },
              activeColor: Colors.white,
              borderColor: Colors.black,
              // you can either pass inactiveColor or gradient for inactive text
              inactiveColor: Colors.green,
              // if both inactiveColor and gradient are passed then gradient will take precedence
              gradient: LinearGradient(
                colors: [Colors.blue, Colors.lightBlueAccent],
                begin: Alignment.bottomLeft,
                end: Alignment.topRight,
              ),
            ),
            SizedBox(height: 12.0),
            Text('Value : $isActive', style: TextStyle(fontSize: 24.0))
          ],
        ),
      ),
    );
  }
}
```
## Screenshots

 ![text_switch](https://media.giphy.com/media/yt9XJG7wrUNsU7bFcI/giphy.gif)


## 👨 Developed By

```
Sampath Narayanan
```

<a href="https://www.linkedin.com/in/narayanan-sampath/"><img src="https://img.shields.io/badge/linkedin-%230077B5.svg?&style=for-the-badge&logo=linkedin&logoColor=white"></a>
<a href="https://medium.com/@sampathnarayanan72"><img src="https://img.shields.io/badge/medium-%2312100E.svg?&style=for-the-badge&logo=medium&logoColor=white"></a>
<a href="https://dev.to/narayanansampath"><img src="https://img.shields.io/badge/DEV.TO-%230A0A0A.svg?&style=for-the-badge&logo=dev-dot-to&logoColor=white"></a>

# 👍 How to Contribute

1. Fork it
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -am 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request

All contributions are highly appreciated!

# 📃 License

   MIT License

   Copyright (c) 2020 Sampath Narayanan

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to deal
   in the Software without restriction, including without limitation the rights
   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
   copies of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included in all
   copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
   SOFTWARE.


## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.dev/).

For help on editing package code, view the [documentation](https://flutter.dev/developing-packages/).


CHANGELOG.md
## [0.0.2] - 19/10/2020

* Readme updates.


lib/text_switch.dart
library text_switch;

import 'package:flutter/material.dart';

class TextSwitch extends StatefulWidget {
  final bool value;
  final ValueChanged<bool>? onChanged;
  final Color activeColor;
  final Color inactiveColor;
  final String activeText;
  final String inactiveText;
  final Color textColor;
  final Gradient? gradient;
  final Color borderColor;
  final String? activeTextSubtitle;
  final String? inActiveTextSubtitle;
  final Widget? leadingActiveWidget;
  final Widget? leadingInactiveWidget;
  final double? height;

  const TextSwitch(
      {Key? key,
      required this.value,
      this.onChanged,
      this.activeColor = Colors.white,
      this.inactiveColor = Colors.grey,
      this.activeText = 'On',
      this.inactiveText = 'Off',
      this.gradient,
      this.activeTextSubtitle,
      this.inActiveTextSubtitle,
      this.leadingActiveWidget,
      this.leadingInactiveWidget,
      this.borderColor = Colors.white,
      this.height = 20,
      this.textColor = Colors.black})
      : super(key: key);

  @override
  _TextSwitchState createState() => _TextSwitchState();
}

class _TextSwitchState extends State<TextSwitch>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late bool value;

  @override
  void initState() {
    super.initState();
    _animationController =
        AnimationController(vsync: this, duration: Duration(milliseconds: 60));
    AlignmentTween(
            begin: widget.value ? Alignment.centerRight : Alignment.centerLeft,
            end: widget.value ? Alignment.centerLeft : Alignment.centerRight)
        .animate(CurvedAnimation(
            parent: _animationController, curve: Curves.linear));
    value = widget.value;
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) {
        return GestureDetector(
          onTap: performAnimation,
          onHorizontalDragEnd: (update) {
            performAnimation();
          },
          child: Container(
            width: double.infinity,
            height: widget.height,
            decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(20.0),
                color: widget.inactiveColor,
                gradient: widget.gradient,
                border: Border.all(color: widget.borderColor)),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: <Widget>[
                Expanded(
                  child: ClipRRect(
                    borderRadius: const BorderRadius.only(
                      topLeft: Radius.circular(10),
                    ),
                    child: Container(
                      height: widget.height,
                      alignment: Alignment.center,
                      decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(20),
                          color: value ? widget.activeColor : Colors.transparent),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                        children: [
                          if(widget.leadingActiveWidget != null)
                          Container(
                            decoration: BoxDecoration(
                              shape: BoxShape.circle,
                              border: Border.all(color: Colors.grey.shade500),
                            ),
                            child: CircleAvatar(
                                backgroundColor: Colors.transparent,
                                child: widget.leadingActiveWidget!
                            ),
                          ),
                          Column(
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: [
                              Text(
                                widget.activeText,
                                textAlign: TextAlign.center,
                                style: TextStyle(
                                    color: widget.textColor,
                                    fontWeight: FontWeight.w900,
                                    fontSize: 12.0),
                              ),
                              if(widget.activeTextSubtitle != null)
                                Text(widget.activeTextSubtitle!,
                                    textAlign: TextAlign.center,
                                    style: Theme.of(context)
                                        .textTheme
                                        .subtitle2!
                                        .copyWith(
                                      color: Colors.grey.shade600,
                                    )),
                            ],
                          ),
                        ],
                      ),
                    ),
                  ),
                ),
                Flexible(
                  child: ClipRRect(
                    borderRadius: BorderRadius.only(
                      topRight: Radius.circular(10),
                    ),
                    child: Container(
                      height: widget.height,
                      alignment: Alignment.center,
                      decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(20),
                          color: value ? Colors.transparent : widget.activeColor),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                        children: [
                          if(widget.leadingActiveWidget != null)
                          Container(
                            decoration: BoxDecoration(
                              shape: BoxShape.circle,
                              border: Border.all(color: Colors.grey.shade500),
                            ),
                            child: CircleAvatar(
                                backgroundColor: Colors.transparent,
                                child: widget.leadingInactiveWidget!
                            ),
                          ),
                          Column(
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: [
                              Text(widget.inactiveText,
                                  textAlign: TextAlign.center,
                                  style: TextStyle(
                                      color: widget.textColor,
                                      fontWeight: FontWeight.w900,
                                      fontSize: 12.0)),
                    if(widget.inActiveTextSubtitle != null)
                Text(widget.inActiveTextSubtitle!,
                    textAlign: TextAlign.center,
                    style: Theme.of(context)
                        .textTheme
                        .subtitle2!
                        .copyWith(
                      color: Colors.grey.shade600,
                    )),

                            ],
                          ),
                        ],
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }

  void performAnimation() {
    if (_animationController.isCompleted) {
      _animationController.reverse();
    } else {
      _animationController.forward();
    }
    setState(() => value = !value);
    final ValueChanged<bool>? onChanged = widget.onChanged;
    if (onChanged != null) {
      widget.value == false ? onChanged(true) : onChanged(false);
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:text_switch/text_switch.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomeScreen(),
    );
  }
}

class HomeScreen extends StatefulWidget {
  @override
  _HomeScreenState createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {

  bool isActive = false;
  bool isIconSwitchActive = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Custom Switch Example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              width: 100,
              child: TextSwitch(
                activeText: 'ON',
                inactiveText: 'OFF',
                value: isActive,
                onChanged: (val) {
                  print('Value: $val');
                  setState(() => isActive = val);
                },
                activeColor: Colors.white,
                borderColor: Colors.black,
                // you can either pass inactiveColor or gradient for inactive text
                inactiveColor: Colors.green,
                // if both inactiveColor and gradient are passed then gradient will take precedence
                gradient: LinearGradient(
                  colors: [Colors.blue, Colors.lightBlueAccent],
                  begin: Alignment.bottomLeft,
                  end: Alignment.topRight,
                ),
              ),
            ),
            SizedBox(height: 12.0),
            Text('Value : $isActive', style: TextStyle(fontSize: 24.0)),
            Container(
              width: double.infinity,
              margin: EdgeInsets.all(20),
              child: TextSwitch(
                activeText: 'ON',
                inactiveText: 'OFF',
                value: isIconSwitchActive,
                onChanged: (val) {
                  print('Value: $val');
                  setState(() => isIconSwitchActive = val);
                },
                leadingActiveWidget: Icon(Icons.lightbulb, color: Colors.black),
                leadingInactiveWidget: Icon(Icons.lightbulb_outline, color: Colors.black),
                height: 70,
                activeTextSubtitle: 'Active subtitle',
                inActiveTextSubtitle: 'Inactive subtitle',
                activeColor: Colors.white,
                borderColor: Colors.black,
                // you can either pass inactiveColor or gradient for inactive text
                inactiveColor: Colors.yellow,
              ),
            ),
            Text('Value : $isIconSwitchActive', style: TextStyle(fontSize: 24.0)),
          ],
        ),
      ),
    );
  }
}
